src/simulation/navigation/pinholeCamera/pinholeCamera.cpp

```diff
@@ -0,0 +1,328 @@
+    this->FOVx = 2 * atan2(this->nxPixel*this->wPixel, 2*this->f);
+    this->FOVy = 2 * atan2(this->nyPixel*this->wPixel, 2*this->f);
+}
+
+/*! Method to add landmarks
+ * @param pos: landmark position in planet-rotating frame
+ * @param normal: landmark surface normal in planet-rotating frame
+ */
+void PinholeCamera::addLandmark(Eigen::Vector3d& pos, Eigen::Vector3d& normal){
+    /* Add the landmark position and surface normal */
+    this->r_LP_P.push_back(pos);
+    this->nL_P.push_back(normal);
+
+    /* Create buffer output messages */
+    Message<AccessMsgPayload> *msg;
+    msg = new Message<AccessMsgPayload>;
+    this->accessOutMsgs.push_back(msg);
+
+    /* Expand the access buffer vectors */
+    AccessMsgPayload accMsg;
+    this->accessMsgBuffer.push_back(accMsg);
+}
+
+/*! Loop through landmarks
+*/
+void PinholeCamera::loopLandmarks()
+{
+    /* Define variables */
+    Eigen::Vector3d rLmk, nLmk;
+    Eigen::Vector2i pLmk;
+    bool flagFOV, flagLighting;
+    
+    /* Preallocate visibility status and landmark pixels */
+    this->isvisibleLmk.setZero(this->n);
+    this->pixelLmk.setZero(this->n, 2);
+    
+    /* Initialize visible landmarks */
+    this->nvisibleLmk = 0;
+    
+    /* Loop through landmarks */
+    for (int i = 0; i < this->n; i++){
+        /* Extract ith landmark position and its normal */
+        rLmk = this->r_LP_P[i];
+        nLmk = this->nL_P[i];
+        
+        /* Compute pixel and check access conditions */
+        pLmk = this->computePixel(rLmk);
+        flagFOV = this->checkFOV(pLmk, rLmk, nLmk);
+        flagLighting = this->checkLighting(nLmk);
+        
+        /* Add pixel */
+        if (flagFOV and flagLighting){
+            this->pixelLmk.row(i) = pLmk;
+            this->isvisibleLmk(i) = 1;
+        }
+    }
+}
+
+
+/*! Check lighting condition of a landmark
+ * @param nLmk: landmark surface normal in planet-rotating frame
+*/
+bool PinholeCamera::checkLighting(Eigen::Vector3d nLmk)
+{
+    /* Compute Sun's incident angle */
+    double angleSun;
+    angleSun = M_PI_2 - safeAcos(nLmk.dot(this->e_SP_P) / (this->e_SP_P.norm()*nLmk.norm()));
+    
+    /* Check if the landmark is illuminated */
+    bool flagLighting;
+    flagLighting = false;
+    if (angleSun > this->maskSun){
+        /* Set lighting condition to true */
+        flagLighting = true;
+    }
+    
+    return flagLighting;
+}
+
+/*! Check if a landmark is within camera field of view
+ * @param pLmk: landmark pixel
+ * @param rLmk: landmark position in planet-rotating frame
+ * @param nLmk: landmark surface normal in planet-rotating frame
+*/
+bool PinholeCamera::checkFOV(Eigen::Vector2i pLmk, Eigen::Vector3d rLmk, Eigen::Vector3d nLmk){
+    /* Compute focal direction in P frame */
+    Eigen::Vector3d eC_P, eC_C;
+    eC_C << 0, 0, 1;
+    eC_P = (this->dcm_CB * this->dcm_BP).transpose()*eC_C;
+    
+    /* Define unit-vector from landmark to spacecraft */
+    Eigen::Vector3d dr, eLmk_P;
```

Comment:
Please ensure all variables in this file have comments that explain what they are.

Reply:
Done